File: fsharp/protein-translation/ProteinTranslation.fs
module ProteinTranslation

let translate =
    function
    | "AUG" -> Some "Methionine"
    | "UUU"
    | "UUC" -> Some "Phenylalanine"
    | "UUA"
    | "UUG" -> Some "Leucine"
    | "UCU"
    | "UCC"
    | "UCA"
    | "UCG" -> Some "Serine"
    | "UAU"
    | "UAC" -> Some "Tyrosine"
    | "UGU"
    | "UGC" -> Some "Cysteine"
    | "UGG" -> Some "Tryptophan"
    | "UAA"
    | "UAG"
    | "UGA" -> None
    | codon -> invalidArg codon "Unknown codon"

let proteins rna =
    Seq.chunkBySize 3 rna
    |> Seq.map System.String
    |> Seq.map translate
    |> Seq.takeWhile Option.isSome
    |> Seq.choose id
    |> Seq.toList

File: scala/protein-translation/src/main/scala/ProteinTranslation.scala
object ProteinTranslation {
  def proteins(seq: String): Seq[String] = {
    seq.grouped(3).map(translate).takeWhile(_ != "STOP").toSeq
  }

  private def translate(codon: String): String = codon match {
    case "AUG" => "Methionine"
    case "UUU" | "UUC" => "Phenylalanine"
    case "UUA" | "UUG" => "Leucine"
    case "UCU" | "UCC" | "UCA" | "UCG" => "Serine"
    case "UAU" | "UAC" => "Tyrosine"
    case "UGU" | "UGC" => "Cysteine"
    case "UGG" => "Tryptophan"
    case "UAA" | "UAG" | "UGA" => "STOP"
    case _ => throw new IllegalArgumentException()
  }
}